### PR TITLE
[T4A3][F12-A3]Gabriel Sim

### PR DESCRIPTION
**Pull request by gabrielsim**
Ready for review

### CODE REVIEW

File: src/seedu/addressbook/data/person/Address.java [hunk 1]
```diff
@@ -6,54 +6,33 @@
...
      * @throws IllegalValueException if given address string is invalid.
      */
     public Address(String address, boolean isPrivate) throws IllegalValueException {
-        String trimmedAddress = address.trim();
-        this.isPrivate = isPrivate;
-        if (!isValidAddress(trimmedAddress)) {
-            throw new IllegalValueException(MESSAGE_ADDRESS_CONSTRAINTS);
-        }
-        this.value = trimmedAddress;
+    	super(validateAddress(address), isPrivate);
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java:22]: nice trick! :+1: does this work without issues? I haven't tried it..

File: src/seedu/addressbook/data/person/Contact.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package seedu.addressbook.data.person;
+
+/**
+ * Represents a Person's contract in the address book.
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Contact.java:4]: Typo detected :-P *contact* not *contract*,
Secondly, no need to mention it as in address book, `Contact` can be used in any application requiring to store the contact information

File: src/seedu/addressbook/data/person/Contact.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+public class Contact {
+	
+    public final String value;
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Contact.java:9]: try to make it private. in general, start off with private members, then use protected or public only if absolutely necessary.

File: src/seedu/addressbook/data/person/Contact.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private boolean isPrivate;
+    
+	public Contact(String value, boolean isPrivate) {
+		this.value = value;
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Contact.java:13]: Notice that your indentation mechanism uses tabs for indentation. 
The coding standards we follow require using 4 spaces for indentation.
Tip: you can set Eclipse to do this for you, check out how!

File: src/seedu/addressbook/data/person/Contact.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    
+    @Override
+    public int hashCode() {
+        return value.hashCode();
+    }
+    
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof Contact // instanceof handles nulls
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Contact.java:30]: by virtue of inheritance, email and address may be validated as the same instance, no?
it is better to leave out the `equals` method in the child class

File: src/seedu/addressbook/data/person/Phone.java [hunk 6]
```diff
@@ -6,54 +6,33 @@
...
-                && this.value.equals(((Phone) other).value)); // state check
-    }
-
-    @Override
-    public int hashCode() {
-        return value.hashCode();
+    public static String validatePhone(String phone) throws IllegalValueException {
+    	String trimmedPhone = phone.trim();
+    	
+        if (!trimmedPhone.matches(PHONE_VALIDATION_REGEX)){
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Phone.java:31]: notice how this chunk of code is repeated in all classes, think how you can push this to the parent class to reduce code duplication.

**Comment by okkhoy** [comment]
@gabrielsim1993 
Some comments added. Please ack & close the PR after reading comments.

**Comment by gabrielsim** [comment]
noted with thanks!

**Pull request closed without merging**